src/dd/Simulation.cpp

```diff
@@ -210,37 +249,42 @@ std::map<std::string, std::size_t> sample(const qc::QuantumComputation& qc,
   return counts;
 }
 
+template <const ApproximationStrategy stgy>
 VectorDD simulate(const qc::QuantumComputation& qc, const VectorDD& in,
-                  Package& dd) {
-  auto permutation = qc.initialLayout;
-  auto e = in;
+                  Package& dd, const Approximation<stgy>& approx) {
+  qc::Permutation permutation = qc.initialLayout;
+  dd::VectorDD out = in;
   for (const auto& op : qc) {
-    // SWAP gates can be executed virtually by changing the permutation
-    if (op->getType() == qc::SWAP && !op->isControlled()) {
-      const auto& targets = op->getTargets();
-      std::swap(permutation.at(targets[0U]), permutation.at(targets[1U]));
-      continue;
+    if (isVirtuallySwappable(*op)) {
+      virtualSwap(*op, permutation);
+    } else {
+      out = applyUnitaryOperation(*op, out, dd, permutation);
+
+      // TODO: this applies approximation after each operation.
+      if constexpr (stgy != None) {
+        applyApproximation(out, approx, dd);
+      }
     }
-
-    e = applyUnitaryOperation(*op, e, dd, permutation);
   }
-  changePermutation(e, permutation, qc.outputPermutation, dd);
-  e = dd.reduceGarbage(e, qc.getGarbage());
+
+  changePermutation(out, permutation, qc.outputPermutation, dd);
+  out = dd.reduceGarbage(out, qc.getGarbage());
 
   // properly account for the global phase of the circuit
-  if (std::abs(qc.getGlobalPhase()) > 0) {
-    // create a temporary copy for reference counting
-    auto oldW = e.w;
-    // adjust for global phase
-    const auto globalPhase = ComplexValue{std::polar(1.0, qc.getGlobalPhase())};
-    e.w = dd.cn.lookup(e.w * globalPhase);
-    // adjust reference count
-    dd.cn.incRef(e.w);
-    dd.cn.decRef(oldW);
+  if (fp phase = qc.getGlobalPhase(); hasGlobalPhase(phase)) {
+    applyGlobalPhase(phase, out, dd);
   }
 
-  return e;
+  return out;
 }
+template VectorDD simulate(const qc::QuantumComputation& qc, const VectorDD& in,
+                           Package& dd, const Approximation<None>& approx);
+template VectorDD simulate(const qc::QuantumComputation& qc, const VectorDD& in,
+                           Package& dd,
+                           const Approximation<FidelityDriven>& approx);
+template VectorDD simulate(const qc::QuantumComputation& qc, const VectorDD& in,
+                           Package& dd,
+                           const Approximation<MemoryDriven>& approx);
```

Comment:
Just a very quick side comment, because this is still early in its development and now's probably the time to make more drastic changes:
I am not the biggest fan of using templates here; especially with the explicit instantiations that become necessary due to code being moved to the source file.
Over the past couple of months, we worked hard to reduce the number of templated classes throughout the DD package as they have become quite a pain.

Couldn't we solve this somehow differently? In principle, any approximation functions takes a DD and produces a DD given certain parameters. Do these functions need state? For certain application strategies probably yes. For certain others, not really.
Maybe we can make something out of that.